File: web/src/lib/quota.ts
// Admin-enforced limits: how many repositories (public / private) and how
// much storage an organization — or a user, across every organization they
// own — may consume. null means unlimited. registryd enforces the same rules
// at push time (see registryd/internal/store/quota.go); this module is the
// web-side twin used by server actions and the admin screens.
import { eq, sql } from "drizzle-orm";
import { db } from "@/db";
import { organizationLimits, userLimits } from "@/db/schema";
import { formatBytes } from "./format";

export interface Usage {
  organizations: number;
  publicRepos: number;
  privateRepos: number;
  storageBytes: number;
}

export interface Limits {
  maxOrganizations: number | null;
  maxPublicRepos: number | null;
  maxPrivateRepos: number | null;
  maxStorageBytes: number | null;
}

export const UNLIMITED: Limits = {
  maxOrganizations: null,
  maxPublicRepos: null,
  maxPrivateRepos: null,
  maxStorageBytes: null,
};

export async function getOrgUsage(orgId: string): Promise<Usage> {
  const { rows } = await db.execute(sql`
    SELECT
      (SELECT count(*)::int FROM repositories WHERE organization_id = ${orgId} AND visibility = 'public') AS public_repos,
      (SELECT count(*)::int FROM repositories WHERE organization_id = ${orgId} AND visibility = 'private') AS private_repos,
      COALESCE((SELECT sum(size)::bigint FROM (
        SELECT DISTINCT b.digest, b.size FROM blobs b
        JOIN repository_blobs rb ON rb.blob_digest = b.digest
        JOIN repositories r ON r.id = rb.repository_id
        WHERE r.organization_id = ${orgId}) t), 0) AS storage_bytes`);
  const r = rows[0];
  return {
    organizations: 1,
    publicRepos: Number(r.public_repos),
    privateRepos: Number(r.private_repos),
    storageBytes: Number(r.storage_bytes),
  };
}

/** Everything in organizations where the user is an owner. */
export async function getUserUsage(userId: string): Promise<Usage> {
  const { rows } = await db.execute(sql`
    WITH owned AS (
      SELECT organization_id FROM member WHERE user_id = ${userId} AND role = 'owner'
    )
    SELECT
      (SELECT count(*)::int FROM owned) AS organizations,
      (SELECT count(*)::int FROM repositories WHERE organization_id IN (SELECT organization_id FROM owned) AND visibility = 'public') AS public_repos,
      (SELECT count(*)::int FROM repositories WHERE organization_id IN (SELECT organization_id FROM owned) AND visibility = 'private') AS private_repos,
      COALESCE((SELECT sum(size)::bigint FROM (
        SELECT DISTINCT r.organization_id, b.digest, b.size FROM blobs b
        JOIN repository_blobs rb ON rb.blob_digest = b.digest
        JOIN repositories r ON r.id = rb.repository_id
        WHERE r.organization_id IN (SELECT organization_id FROM owned)) t), 0) AS storage_bytes`);
  const r = rows[0];
  return {
    organizations: Number(r.organizations),
    publicRepos: Number(r.public_repos),
    privateRepos: Number(r.private_repos),
    storageBytes: Number(r.storage_bytes),
  };
}

export async function getOrgLimits(orgId: string): Promise<Limits> {
  const row = await db.query.organizationLimits.findFirst({
    where: eq(organizationLimits.organizationId, orgId),
  });
  return row ? { maxOrganizations: null, ...pick(row) } : UNLIMITED;
}

export async function getUserLimits(userId: string): Promise<Limits> {
  const row = await db.query.userLimits.findFirst({ where: eq(userLimits.userId, userId) });
  return row ? { maxOrganizations: row.maxOrganizations, ...pick(row) } : UNLIMITED;
}

function pick(row: {
  maxPublicRepos: number | null;
  maxPrivateRepos: number | null;
  maxStorageBytes: number | null;
}) {
  return {
    maxPublicRepos: row.maxPublicRepos,
    maxPrivateRepos: row.maxPrivateRepos,
    maxStorageBytes: row.maxStorageBytes,
  };
}

async function orgOwnerIds(orgId: string): Promise<string[]> {
  const { rows } = await db.execute(sql`
    SELECT user_id FROM member WHERE organization_id = ${orgId} AND role = 'owner'`);
  return rows.map((r) => r.user_id as string);
}

/**
 * Can one more repository of this visibility be created in the org?
 * Returns a human-readable reason when not.
 */
export async function checkRepoQuota(
  orgId: string,
  visibility: "public" | "private",
  orgLabel = "this organization",
  /** Repositories of this visibility already promised to the org but not stored yet (bulk preview). */
  pending = 0,
): Promise<string | null> {
  const key = visibility === "public" ? "maxPublicRepos" : "maxPrivateRepos";
  const usageKey = visibility === "public" ? "publicRepos" : "privateRepos";

  const [orgLimit, orgUsage] = await Promise.all([getOrgLimits(orgId), getOrgUsage(orgId)]);
  if (orgLimit[key] !== null && orgUsage[usageKey] + pending >= orgLimit[key]) {
    return `${orgLabel} has reached its limit of ${orgLimit[key]} ${visibility} repositories.`;
  }
  for (const ownerId of await orgOwnerIds(orgId)) {
    const [limit, usage] = await Promise.all([getUserLimits(ownerId), getUserUsage(ownerId)]);
    if (limit[key] !== null && usage[usageKey] + pending >= limit[key]) {
      return `The organization owner's account has reached its limit of ${limit[key]} ${visibility} repositories.`;
    }
  }
  return null;
}

/** Would adding `additionalBytes` to the org exceed any storage limit? */
export async function checkStorageQuota(orgId: string, additionalBytes: number): Promise<string | null> {
  const [orgLimit, orgUsage] = await Promise.all([getOrgLimits(orgId), getOrgUsage(orgId)]);
  if (orgLimit.maxStorageBytes !== null && orgUsage.storageBytes + additionalBytes > orgLimit.maxStorageBytes) {
    return `Storage limit reached: ${formatBytes(orgUsage.storageBytes)} of ${formatBytes(orgLimit.maxStorageBytes)} used.`;
  }
  for (const ownerId of await orgOwnerIds(orgId)) {
    const [limit, usage] = await Promise.all([getUserLimits(ownerId), getUserUsage(ownerId)]);
    if (limit.maxStorageBytes !== null && usage.storageBytes + additionalBytes > limit.maxStorageBytes) {
      return `The owner's account storage limit is reached: ${formatBytes(usage.storageBytes)} of ${formatBytes(limit.maxStorageBytes)}.`;
    }
  }
  return null;
}

/** May the user create another organization? */
export async function checkOrgCreationQuota(userId: string): Promise<string | null> {
  const [limit, usage] = await Promise.all([getUserLimits(userId), getUserUsage(userId)]);
  if (limit.maxOrganizations !== null && usage.organizations >= limit.maxOrganizations) {
    return `You have reached your limit of ${limit.maxOrganizations} organizations.`;
  }
  return null;
}

/** Parse a limits form: empty string → unlimited (null). */
export function parseLimitField(value: FormDataEntryValue | null): number | null {
  const s = String(value ?? "").trim();
  if (s === "") return null;
  const n = Number(s);
  if (!Number.isFinite(n) || n < 0) return null;
  return Math.floor(n);
}

/** Storage entered in GiB on the admin forms. */
export function parseStorageGiB(value: FormDataEntryValue | null): number | null {
  const s = String(value ?? "").trim();
  if (s === "") return null;
  const n = Number(s);
  if (!Number.isFinite(n) || n < 0) return null;
  return Math.round(n * 1024 * 1024 * 1024);
}

File: registryd/internal/store/quota.go
package store

import (
	"context"
	"errors"
	"fmt"
)

// QuotaError describes a limit an admin configured that a push would exceed.
type QuotaError struct {
	Scope string // "organization" or "owner"
	Kind  string // "storage", "private repositories", "public repositories"
	Used  int64
	Limit int64
}

func (e *QuotaError) Error() string {
	if e.Kind == "storage" {
		return fmt.Sprintf("%s storage quota exceeded: %s of %s used", e.Scope, humanBytes(e.Used), humanBytes(e.Limit))
	}
	return fmt.Sprintf("%s quota exceeded: %d of %d %s used", e.Scope, e.Used, e.Limit, e.Kind)
}

// IsQuotaError reports whether err is a QuotaError.
func IsQuotaError(err error) bool {
	var qe *QuotaError
	return errors.As(err, &qe)
}

func humanBytes(n int64) string {
	const unit = 1024
	if n < unit {
		return fmt.Sprintf("%d B", n)
	}
	div, exp := int64(unit), 0
	for m := n / unit; m >= unit; m /= unit {
		div *= unit
		exp++
	}
	return fmt.Sprintf("%.1f %ciB", float64(n)/float64(div), "KMGTPE"[exp])
}

// orgStorageUsed sums the distinct blobs linked to the organization's repos.
const orgStorageUsedSQL = `
	SELECT COALESCE(sum(size), 0)::bigint FROM (
		SELECT DISTINCT b.digest, b.size FROM blobs b
		JOIN repository_blobs rb ON rb.blob_digest = b.digest
		JOIN repositories r ON r.id = rb.repository_id
		WHERE r.organization_id = $1) t`

// ownerStorageUsedSQL sums usage across the organizations the user owns
// that have no storage limit of their own: an organization with its own
// limit is governed by that limit alone (see web/src/lib/quota.ts).
const ownerStorageUsedSQL = `
	SELECT COALESCE(sum(size), 0)::bigint FROM (
		SELECT DISTINCT r.organization_id, b.digest, b.size FROM blobs b
		JOIN repository_blobs rb ON rb.blob_digest = b.digest
		JOIN repositories r ON r.id = rb.repository_id
		JOIN member m ON m.organization_id = r.organization_id
		LEFT JOIN organization_limits ol ON ol.organization_id = r.organization_id
		WHERE m.user_id = $1 AND m.role = 'owner' AND ol.max_storage_bytes IS NULL) t`

// CheckStorageQuota fails with a *QuotaError when adding `additional` bytes
// to the organization would exceed the org's own limit or, when it has
// none, any owner's account limit.
func (s *Store) CheckStorageQuota(ctx context.Context, orgID string, additional int64) error {
	var limit *int64
	err := s.pool.QueryRow(ctx, `SELECT max_storage_bytes FROM organization_limits WHERE organization_id = $1`, orgID).Scan(&limit)
	if err != nil && !isNoRows(err) {
		return err
	}
	if limit != nil {
		var used int64
		if err := s.pool.QueryRow(ctx, orgStorageUsedSQL, orgID).Scan(&used); err != nil {
			return err
		}
		if used+additional > *limit {
			return &QuotaError{Scope: "organization", Kind: "storage", Used: used, Limit: *limit}
		}
		// The organization's own limit governs it; the owners' accounts are not consulted.
		return nil
	}

	rows, err := s.pool.Query(ctx, `
		SELECT ul.user_id, ul.max_storage_bytes FROM member m
		JOIN user_limits ul ON ul.user_id = m.user_id
		WHERE m.organization_id = $1 AND m.role = 'owner' AND ul.max_storage_bytes IS NOT NULL`, orgID)
	if err != nil {
		return err
	}
	defer rows.Close()
	type owner struct {
		id    string
		limit int64
	}
	var owners []owner
	for rows.Next() {
		var o owner
		if err := rows.Scan(&o.id, &o.limit); err != nil {
			return err
		}
		owners = append(owners, o)
	}
	if err := rows.Err(); err != nil {
		return err
	}
	for _, o := range owners {
		var used int64
		if err := s.pool.QueryRow(ctx, ownerStorageUsedSQL, o.id).Scan(&used); err != nil {
			return err
		}
		if used+additional > o.limit {
			return &QuotaError{Scope: "owner", Kind: "storage", Used: used, Limit: o.limit}
		}
	}
	return nil
}

// CheckRepositoryQuota fails when the organization (or one of its owners)
// may not create another repository of the given visibility.
func (s *Store) CheckRepositoryQuota(ctx context.Context, orgID, visibility string) error {
	column := "max_private_repos"
	kind := "private repositories"
	if visibility == "public" {
		column = "max_public_repos"
		kind = "public repositories"
	}

	var limit *int64
	err := s.pool.QueryRow(ctx, `SELECT `+column+` FROM organization_limits WHERE organization_id = $1`, orgID).Scan(&limit)
	if err != nil && !isNoRows(err) {
		return err
	}
	if limit != nil {
		var used int64
		if err := s.pool.QueryRow(ctx, `
			SELECT count(*) FROM repositories WHERE organization_id = $1 AND visibility = $2`, orgID, visibility).Scan(&used); err != nil {
			return err
		}
		if used >= *limit {
			return &QuotaError{Scope: "organization", Kind: kind, Used: used, Limit: *limit}
		}
		// The organization's own limit governs it; the owners' accounts are not consulted.
		return nil
	}

	rows, err := s.pool.Query(ctx, `
		SELECT ul.user_id, ul.`+column+` FROM member m
		JOIN user_limits ul ON ul.user_id = m.user_id
		WHERE m.organization_id = $1 AND m.role = 'owner' AND ul.`+column+` IS NOT NULL`, orgID)
	if err != nil {
		return err
	}
	defer rows.Close()
	type owner struct {
		id    string
		limit int64
	}
	var owners []owner
	for rows.Next() {
		var o owner
		if err := rows.Scan(&o.id, &o.limit); err != nil {
			return err
		}
		owners = append(owners, o)
	}
	if err := rows.Err(); err != nil {
		return err
	}
	for _, o := range owners {
		var used int64
		// Organizations with their own limit of this kind are outside the owner's pool.
		if err := s.pool.QueryRow(ctx, `
			SELECT count(*) FROM repositories r
			JOIN member m ON m.organization_id = r.organization_id
			LEFT JOIN organization_limits ol ON ol.organization_id = r.organization_id
			WHERE m.user_id = $1 AND m.role = 'owner' AND r.visibility = $2 AND ol.`+column+` IS NULL`, o.id, visibility).Scan(&used); err != nil {
			return err
		}
		if used >= o.limit {
			return &QuotaError{Scope: "owner", Kind: kind, Used: used, Limit: o.limit}
		}
	}
	return nil
}

// OrgHasBlob reports whether the blob is already linked to any repository of
// the organization (and therefore already counted against its storage).
func (s *Store) OrgHasBlob(ctx context.Context, orgID, digest string) (bool, error) {
	var one int
	err := s.pool.QueryRow(ctx, `
		SELECT 1 FROM repository_blobs rb JOIN repositories r ON r.id = rb.repository_id
		WHERE r.organization_id = $1 AND rb.blob_digest = $2 LIMIT 1`, orgID, digest).Scan(&one)
	if isNoRows(err) {
		return false, nil
	}
	return err == nil, err
}
